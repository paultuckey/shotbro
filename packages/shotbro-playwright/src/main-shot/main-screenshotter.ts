import type {BrowserContext, Page} from "@playwright/test";
import {ShotBroBox, ShotBroFocus, ShotBroInput} from "../annotate/annotate-types";
import {ShapePosition} from "../annotate/shape/shape-types";
import {atToLocatorStr} from "../annotate/shape/shape-locator";
import {stringifySbBb} from "../annotate/box-utils";
import {generateHtmlForOverlayString} from "../annotate/overlay/overlay-html-generator";

const MAIN_CAPTURE_HEIGHT_LIMIT = 4000;
const MAIN_CAPTURE_WIDTH_LIMIT = 4000;

export type InputPositions = {
    focusBoxPosition?: ShotBroBox,
    shapePositions: ShotBroBox[]
}

export async function generateMainScreenshot(page: Page, screenshotPath: string) {
    const rootEl = await page.locator('html');
    const rootBb = await rootEl.boundingBox()
    if (!rootBb) throw new Error('Could not get size of HTML element');
    await page.screenshot({
        path: screenshotPath,
        fullPage: true,
        animations: 'disabled',
        clip: {
            x: rootBb.x, y: rootBb.y,
            width: Math.min(rootBb.width, MAIN_CAPTURE_WIDTH_LIMIT),
            height: Math.min(rootBb.height, MAIN_CAPTURE_HEIGHT_LIMIT),
        },
        type: 'png',
        scale: 'device' // we want to get the best quality possible
    });
}


export async function findPositions(page: Page, input: ShotBroInput): Promise<InputPositions> {
    //const inputPositionsOuter = await page.evaluate(async (inputInner) => {
    // otherwise we are very specifically not touching playwright's setup as the
    // test may have scrolled something (like a div within a larger div)

    // suggest: optionally ensure the page is scrolled to the top
    // window.scroll({top: 0, left: 0, behavior: 'auto'});

    //let page = null;
    async function calculateShapePosition(page: Page, shapePosition: ShapePosition) {
        return await calculateFocusBox(page, shapePosition);
    }

    /**
     * Calculate the position of the box the user selected.  With no padding or other modifiers.
     */
    async function calculateFocusBox(page: Page, focusBox: ShotBroFocus): Promise<ShotBroBox> {
        let atBox;
        if (focusBox?.atPos) {
            atBox = focusBox?.atPos;
        } else {
            let locatorStr = atToLocatorStr(focusBox);
            const atLocator = page.locator(locatorStr);
            let sbBb;
            if (atLocator) {
                sbBb = await atLocator.boundingBox();
                if (sbBb) {
                    atBox = {x: sbBb.x, y: sbBb.y, w: sbBb.width, h: sbBb.height};
                } else {
                    console.log(`Unable to find data-sb-bb attribute for ${locatorStr}`);
                }
            } else {
                console.log('Unable to find a locator node main screenshot')
            }
            if (!atBox) {
                console.log('unable to calculate focus box, using default')
                atBox = {w: 1024, h: 768, x: 0, y: 0};
            }
        }
        console.log(`calculated focusBox ${stringifySbBb(atBox)}`)
        return atBox;
    }

    let focusBoxPosition: ShotBroBox | undefined = undefined;
    let shapePositions: ShotBroBox[] = [];
    if (input.focus && input.focus.at) {
        focusBoxPosition = await calculateFocusBox(page, input.focus);
    }
    if (input.shapes) {
        for (let i = 0; i < input.shapes.length; i++) {
            let shape = input.shapes[i];
            let shapePos: ShotBroBox | null = null;
            if (shape.arrow) {
                shapePos = await calculateShapePosition(page, shape.arrow);
            }
            if (shape.box) {
                shapePos = await calculateShapePosition(page, shape.box);
            }
            if (shape.circle) {
                shapePos = await calculateShapePosition(page, shape.circle);
            }
            if (shape.text) {
                shapePos = await calculateShapePosition(page, shape.text);
            }
            if (shapePos != null) {
                shapePositions[i] = shapePos;
            }
        }
    }
    return {
        focusBoxPosition, shapePositions
    };
}

export async function shotBroPlaywrightAnnotate(context: BrowserContext, input: ShotBroInput, inputPositions: InputPositions, focusPngPath: string) {
    let html = await generateHtmlForOverlayString(input, inputPositions);
    let page = await context.newPage();
    await page.goto('data:text/html,' + html);
    await page.waitForTimeout(1000);
    await page.screenshot({path: focusPngPath});
    await page.close();
}
