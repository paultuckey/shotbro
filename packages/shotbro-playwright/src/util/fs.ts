import {CliLog} from "./log";
import fs from "node:fs/promises";
import path from "node:path";
import os from "node:os";

export async function cleanupDir(log: CliLog, dirToClean: string) {
    if (!dirToClean) return;
    log.debug(`Cleaning ${dirToClean}`)
    try {
        await fs.rm(dirToClean, {recursive: true});
    } catch (e) {
        log.warn(`An error has occurred while cleaning up ${dirToClean}. Error: ${e}`);
    }
}

export async function prepareDir(log: CliLog, outDir: string) {
    if (!outDir) {
        outDir = await fs.mkdtemp(path.join(os.tmpdir(), 'ShotBro'));
        log.debug(`dir is not set, using temp dir ${outDir}`);
    } else {
        try {
            await fs.access(outDir);
            log.debug(`dir access ok ${outDir}`)
        } catch (e) {
            log.debug(`dir access fail, creating ${outDir}`)
            await fs.mkdir(outDir, {recursive: true});
        }
    }
    return path.resolve(outDir);
}
