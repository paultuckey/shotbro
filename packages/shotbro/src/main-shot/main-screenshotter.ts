import type {Page} from "playwright";
import * as fs from "fs";

export async function generateMainScreenshot(page: Page, screenshotPath: string, htmlContentPath: string) {
  await page.screenshot({
    path: screenshotPath,
    fullPage: true,
    animations: 'disabled',
  })

  await page.evaluate(async () => {
    // otherwise we are very specifically not touching playwright's setup as the
    // test may have scrolled something (like a div within a larger div)

    // sugges: optionally ensure the page is scrolled to the top
    // window.scroll({top: 0, left: 0, behavior: 'auto'});

    const addSbBb = async function (node) {
      // https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect
      const bb = await node.getBoundingClientRect();
      // data-sb-bb = Shot Bro Bounding Box / in css style key:value;...
      node.setAttribute('data-sb-bb', `x:${bb.x};y:${bb.y};w:${bb.width};h:${bb.height};`)
    };
    await addSbBb(document.body);
    document.body.querySelectorAll('*').forEach(addSbBb);
  });
  const content = await page.content();
  fs.writeFileSync(htmlContentPath, content);
}
