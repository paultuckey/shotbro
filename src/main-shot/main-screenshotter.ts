import type {Page} from "playwright";
import * as fs from "fs";

const MAIN_CAPTURE_HEIGHT_LIMIT = 4000;

export async function generateMainScreenshot(page: Page, htmlContentPath: string, screenshotPath: string) {
  const body  = await page.locator('body');
  const bodyBb = await body.boundingBox()
  if (!bodyBb) throw new Error('Could not get size of HTML body');
  await page.screenshot({
    path: screenshotPath,
    fullPage: true,
    animations: 'disabled',
    clip: {
      x: bodyBb.x, y: bodyBb.y,
      width: bodyBb.width, // todo: future, handle very wide also
      height: bodyBb.height > 4000 ? MAIN_CAPTURE_HEIGHT_LIMIT: bodyBb.height
    }
  });

  await page.evaluate(async () => {
    // otherwise we are very specifically not touching playwright's setup as the
    // test may have scrolled something (like a div within a larger div)

    // sugges: optionally ensure the page is scrolled to the top
    // window.scroll({top: 0, left: 0, behavior: 'auto'});

    const addSbBb = async function (el: Element) {
      // https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect
      const bb = await el.getBoundingClientRect();
      // data-sb-bb = Shot Bro Bounding Box / in css style key:value;...
      el.setAttribute('data-sb-bb', `x:${bb.x};y:${bb.y};w:${bb.width};h:${bb.height};`)
    };
    await addSbBb(document.body);
    document.body.querySelectorAll('*').forEach(addSbBb);
  });
  const content = await page.content();
  fs.writeFileSync(htmlContentPath, content);
}
