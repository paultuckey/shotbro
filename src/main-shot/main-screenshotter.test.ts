import {chromium, Browser, Page} from 'playwright';
import * as path from "path";
import * as fs from "fs";
import {generateMainScreenshot} from "./main-screenshotter";
import {COMPARE_DIR_NAME, expectImageToMatchBaseline, SNAPSHOTS_DIR_NAME} from "../test/test-utils";
import {PNG} from 'pngjs';

describe('Main screenshotter', () => {
  describe('Box', () => {

    let browser: Browser;
    let page: Page;

    beforeAll(async () => browser = await chromium.launch());
    afterAll(async () => await browser.close());
    beforeEach(async () => {
      page = await browser.newPage();
      await page.setViewportSize({width: 320, height: 320});
      await page.goto(`file:${path.join(__dirname, '..', 'test', 'boxes.html')}`);
    });
    afterEach(async () => await page.close());

    test('when default used, box should match whole page html', async () => {
      const pngPath = path.join(__dirname, SNAPSHOTS_DIR_NAME, COMPARE_DIR_NAME, 'main-default.png');
      const htmlPath = path.join(__dirname, SNAPSHOTS_DIR_NAME, COMPARE_DIR_NAME, 'main-default.html');
      await generateMainScreenshot(page, htmlPath, pngPath);
      const outHtml = fs.readFileSync(htmlPath, 'utf-8');
      expect(outHtml).toMatchSnapshot('main-default');
    })

    test('when default used, box should match whole page png', async () => {
      const pngPath = path.join(__dirname, SNAPSHOTS_DIR_NAME, COMPARE_DIR_NAME, 'main-default.png');
      const htmlPath = path.join(__dirname, SNAPSHOTS_DIR_NAME, COMPARE_DIR_NAME, 'main-default.html');
      await generateMainScreenshot(page, htmlPath, pngPath);
      await expectImageToMatchBaseline(pngPath);
    })

  });

  describe('Box extra long', () => {

    let browser: Browser;
    let page: Page;

    beforeAll(async () => browser = await chromium.launch());
    afterAll(async () => await browser.close());
    beforeEach(async () => {
      page = await browser.newPage();
      await page.setViewportSize({width: 320, height: 320});
      await page.goto(`file:${path.join(__dirname, '..', 'test', 'boxes-extra-long.html')}`);
    });
    afterEach(async () => await page.close());

    test('when page extra long capture should be limited', async () => {
      const pngPath = path.join(__dirname, SNAPSHOTS_DIR_NAME, COMPARE_DIR_NAME, 'main-limited.png');
      const htmlPath = path.join(__dirname, SNAPSHOTS_DIR_NAME, COMPARE_DIR_NAME, 'main-limited.html');
      await generateMainScreenshot(page, htmlPath, pngPath);

      expect(PNG.sync.read(fs.readFileSync(pngPath)).height).toBe(4000);
      await expectImageToMatchBaseline(pngPath);
    })


  });
});