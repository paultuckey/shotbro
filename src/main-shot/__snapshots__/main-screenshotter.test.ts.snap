// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main screenshotter Box when default used, box should match whole page html: main-default 1`] = `
"<html lang="en" data-sb-bb="x:0;y:0;w:320;h:630;"><head>
    <title>Test</title>
    <!-- we need to ensure there are minimal cross OS differences or else matching basline images won't work -->
    <link href="../../node_modules/modern-normalize/modern-normalize.css" rel="stylesheet">
    <style>
        body {
            font-family: Inter, sans-serif;
            background-color: whitesmoke;
        }
    </style>
</head>
<body class="p-4" style="background-color: whitesmoke;display: flex;flex-wrap: wrap;align-content: flex-start;" data-sb-bb="x:0;y:0;w:320;h:630;">

<div id="box1" style="background-color: red;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:10;y:10;w:50;h:50;"></div>
<div id="box2" style="background-color: orange;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:80;y:10;w:50;h:50;"></div>
<div id="box3" style="background-color: yellow;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:150;y:10;w:50;h:50;"></div>
<div style="background-color: green;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:220;y:10;w:50;h:50;"></div>
<div style="background-color: blue;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:10;y:80;w:50;h:50;"></div>
<div style="background-color: indigo;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:80;y:80;w:50;h:50;"></div>
<div style="background-color: violet;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:150;y:80;w:50;h:50;"></div>

<div id="box8" style="background-color: red;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:220;y:80;w:50;h:50;"></div>
<div style="background-color: orange;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:10;y:150;w:50;h:50;"></div>
<div style="background-color: yellow;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:80;y:150;w:50;h:50;"></div>
<div style="background-color: green;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:150;y:150;w:50;h:50;"></div>
<div style="background-color: blue;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:220;y:150;w:50;h:50;"></div>
<div style="background-color: indigo;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:10;y:220;w:50;h:50;"></div>
<div style="background-color: violet;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:80;y:220;w:50;h:50;"></div>

<div id="box15" style="background-color: red;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:150;y:220;w:50;h:50;"></div>
<div style="background-color: orange;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:220;y:220;w:50;h:50;"></div>
<div style="background-color: yellow;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:10;y:290;w:50;h:50;"></div>
<div style="background-color: green;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:80;y:290;w:50;h:50;"></div>
<div style="background-color: blue;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:150;y:290;w:50;h:50;"></div>
<div style="background-color: indigo;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:220;y:290;w:50;h:50;"></div>
<div style="background-color: violet;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:10;y:360;w:50;h:50;"></div>

<div id="box22" style="background-color: red;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:80;y:360;w:50;h:50;"></div>
<div style="background-color: orange;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:150;y:360;w:50;h:50;"></div>
<div style="background-color: yellow;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:220;y:360;w:50;h:50;"></div>
<div style="background-color: green;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:10;y:430;w:50;h:50;"></div>
<div style="background-color: blue;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:80;y:430;w:50;h:50;"></div>
<div style="background-color: indigo;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:150;y:430;w:50;h:50;"></div>
<div style="background-color: violet;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:220;y:430;w:50;h:50;"></div>

<div style="background-color: red;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:10;y:500;w:50;h:50;"></div>
<div style="background-color: orange;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:80;y:500;w:50;h:50;"></div>
<div style="background-color: yellow;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:150;y:500;w:50;h:50;"></div>
<div style="background-color: green;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:220;y:500;w:50;h:50;"></div>
<div style="background-color: blue;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:10;y:570;w:50;h:50;"></div>
<div style="background-color: indigo;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:80;y:570;w:50;h:50;"></div>
<div style="background-color: violet;width: 50px;height: 50px;margin: 10px;" data-sb-bb="x:150;y:570;w:50;h:50;"></div>



</body></html>"
`;
